include: "rules/common.smk"
include: "rules/trimmomatic.smk"
include: "rules/subsample.smk"
include: "rules/quality.smk"
include: "rules/genome_preparation.smk"
include: "rules/align.smk"
include: "rules/methylation_extraction.smk"
include: "rules/dmrcaller.smk"

report: "report/workflow.rst"

# Allow users to fix the underlying OS via singularity.
singularity: "docker://continuumio/miniconda3:4.4.10"


def all_input(wildcards):
	"""
	Function defining all requested inputs for the rule all (below).
	"""

	wanted_input = []


	# Quality report

	wanted_input.extend(
			expand(
				[
					"results/quality/multiqc/{unit.sample}{unit.lane}{unit.techrep}-{unit.biorep}/multiqc_report.html",

				],
				unit=get_unit(),

			)
		)
	# Alignments 


	wanted_input.extend(
		expand(
			[
				"results/alignment/{unit.sample}{unit.techrep}-{unit.biorep}/{unit.sample}{unit.techrep}-{unit.biorep}_alignment.deduplicated.sam",
			] ,
			unit=get_merged(),
		)
	)

	# Bismark methylation extraction report 
	
	wanted_input.extend(
		expand(
			[
				"results/methylation_extraction/{unit.sample}{unit.techrep}-{unit.biorep}/{unit.sample}{unit.techrep}-{unit.biorep}_alignment.deduplicated.CX_report.txt"
			] ,
			unit=get_merged(),
		)
	)
	wanted_input.extend(
		expand(
			[
				"results/methylation_calling/control_vs_{sample}{techrep}.Rdata"
			] ,
			sample=samples['sample'],
			techrep=['1']
		)
	)

	return wanted_input

rule all:
	input: all_input

